### PR TITLE
enable ASO in creating configMap for OIDC Issuer URL

### LINKED ISSUE #4298: Export OIDC issuer URL to ConfigMap
/kind feature
/area managedclusters

**Describe the solution you'd like**

Since CAPZ `AzureManagedControlPlane` is now backed by ASO `ManagedCluster`, we should export the cluster's OIDC issuer URL to a ConfigMap by using `.spec.operatorSpec.configMaps.oidcIssuerProfile` ([spec](https://azure.github.io/azure-service-operator/reference/containerservice/v1api20230201/#containerservice.azure.com/v1api20230201.ManagedClusterOperatorConfigMaps)).

This would be very helpful when using ASO to provision Azure resources such as `FederatedIdentityCredential`. This resource can read the cluster OIDC issuer URL via [ConfigMap sharing](https://azure.github.io/azure-service-operator/guide/configmaps/), as of https://github.com/Azure/azure-service-operator/pull/3125 and https://github.com/Azure/azure-service-operator/pull/3126.

**Additional information**

The ASO `ManagedCluster` created by CAPZ uses API version `containerservice.azure.com/v1api20210501`. I think we would need to be on at least `containerservice.azure.com/v1api20230201` to do this.

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- Enables ASO with OIDC issuer configMap details so that ASO can create the configMap.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4298 

**Special notes for your reviewer**:
- [x] ~Notice that I had _enabled OIDC Issuer Profile_ at `AzureManagedControlPlane` level. I still need to probe if ` OIDC Issuer Profile` should be enabled to begin with.~
  - Update: `AzureManagedControlPlane.Spec.oidcIssuerProfile.enabled` must be enabled. Refer https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4551#issuecomment-1935085979

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
enable ASO in creating configMap for OIDC Issuer URL
```

### CODE REVIEW

File: azure/services/managedclusters/spec.go [hunk 1]
```diff
@@ -724,6 +731,10 @@ func adminKubeconfigSecretName(clusterName string) string {
 	return secret.Name(clusterName+"-aso", secret.Kubeconfig)
 }
 
+func oidcIssuerURLConfigMapName(clusterName string) string {
```
Comment (jackfrancis) [inline azure/services/managedclusters/spec.go:734]: let's set up a simple UT for this new func

File: azure/services/managedclusters/spec.go [hunk 2]
```diff
@@ -487,7 +487,14 @@ func (s *ManagedClusterSpec) Parameters(ctx context.Context, existing *asocontai
 				Key:  secret.KubeconfigDataName,
 			},
 		},
+		ConfigMaps: &asocontainerservicev1.ManagedClusterOperatorConfigMaps{
+			OIDCIssuerProfile: &genruntime.ConfigMapDestination{
+				Name: oidcIssuerURLConfigMapName(s.ClusterName),
+				Key:  "oidc-issuer-profile-url",
```
Comment (jackfrancis) [inline azure/services/managedclusters/spec.go:493]: `"oidc-issuer-profile-url"` should be a const

File: azure/services/managedclusters/spec.go (hunk 1 above)
Comment (nawazkh) [reply azure/services/managedclusters/spec.go:734]: added via fe9c588ee0b05ef1f32b897f23b532fe76994260

File: azure/services/managedclusters/spec.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nawazkh) [reply azure/services/managedclusters/spec.go]: added via fe9c588ee0b05ef1f32b897f23b532fe76994260

File: azure/services/managedclusters/spec_test.go [hunk 3]
```diff
@@ -236,6 +236,12 @@ func TestParameters(t *testing.T) {
 							Key:  secret.KubeconfigDataName,
 						},
 					},
+					ConfigMaps: &asocontainerservicev1.ManagedClusterOperatorConfigMaps{
+						OIDCIssuerProfile: &genruntime.ConfigMapDestination{
+							Name: "cluster-aso-oidc-issuer-profile",
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply azure/services/managedclusters/spec_test.go]: Was following the above code, made the changes via fe9c588ee0b05ef1f32b897f23b532fe76994260

File: azure/services/managedclusters/spec.go (hunk 2 above)
Comment (nawazkh) [reply azure/services/managedclusters/spec.go]: added via fe9c588ee0b05ef1f32b897f23b532fe76994260

**Comment by nawazkh** [comment]
Squashed and rebased, ready for review!

**Comment by nawazkh** [comment]
Running a test to check if there's a co-relation between OIDC Issuer configMap being created and Managed Cluster's `OIDCIssuerProfile.Enabled`.

/hold

File: templates/flavors/aks/cluster-template.yaml [hunk 4]
```diff
@@ -33,6 +33,8 @@ spec:
   location: "${AZURE_LOCATION}"
   sshPublicKey: ${AZURE_SSH_PUBLIC_KEY_B64:=""}
   version: "${KUBERNETES_VERSION}"
+  oidcIssuerProfile:
+    enabled: true
```
Comment (nawazkh) [inline templates/flavors/aks/cluster-template.yaml:37]: Do I need to enable this? Running a sanity test locally. BRB

Comment (jackfrancis) [reply templates/flavors/aks/cluster-template.yaml:37]: ya

File: azure/services/managedclusters/spec.go [hunk 5]
```diff
@@ -487,7 +487,14 @@ func (s *ManagedClusterSpec) Parameters(ctx context.Context, existing *asocontai
 				Key:  secret.KubeconfigDataName,
 			},
 		},
+		ConfigMaps: &asocontainerservicev1.ManagedClusterOperatorConfigMaps{
```
Comment (jackfrancis) [inline azure/services/managedclusters/spec.go:490]: I think we want to move this entire block down into the block at L591

Something like 

```go
        if s.OIDCIssuerProfile != nil {
		managedCluster.Spec.OidcIssuerProfile = &asocontainerservicev1.ManagedClusterOIDCIssuerProfile{
		    Enabled: s.OIDCIssuerProfile.Enabled,
		}
                if s.OIDCIssuerProfile.Enabled {
                    managedCluster.Spec.OperatorSpec.ConfigMaps = &asocontainerservicev1.ManagedClusterOperatorConfigMaps{
                        OIDCIssuerProfile: &genruntime.ConfigMapDestination{
			    Name: oidcIssuerURLConfigMapName(s.ClusterName),
			    Key:  oidcIssuerProfileURL,
			},
                    }
                }
	}
```

This way we don't ship a OIDCIssuer URL config when oidc is disabled.

**Comment by jackfrancis** [comment]
Let's add a brief note in `docs/book/src/topics/managedcluster.md` under the "Security Profile for AKS clusters" heading that when you configure `oidcIssuerProfile.enabled` to true on your `AzureManagedControlPlane` resource, you will have a configmap named "<cluster name>-aso-oidc-issuer-profile" in the same namespace as your `AzureManagedControlPlane` that contains the oidc URL under the `oidc-issuer-profile-url` key.

File: templates/flavors/aks/cluster-template.yaml (hunk 4 above)
Comment (nawazkh) [reply templates/flavors/aks/cluster-template.yaml:37]: Tested it locally. I must have `AzureManagedControlPlane.Spec.oidcIssuerProfile.enabled:true` in order to have OIDC ConfigMap created. 
``` 
❯ kg secrets -A -w
NAMESPACE                           NAME                                              TYPE                            DATA   AGE
.
default                             aks-28954-aso-kubeconfig                          Opaque                          1      7m23s
default                             aks-28954-aso-secret                              Opaque                          4      14m
.
default                             aks-6439-aso-kubeconfig                           Opaque                          1      31m
default                             aks-6439-aso-secret                               Opaque                          4      39m
.
```

```
❯ kg configmaps -A
NAMESPACE                           NAME                                                   DATA   AGE
.
default                             aks-6439-aso-oidc-issuer-profile                       1      31m
.
```

Context: `aks-6439` has  `AzureManagedControlPlane.Spec.oidcIssuerProfile.enabled:true`  and `aks-28954` has  `AzureManagedControlPlane.Spec.oidcIssuerProfile` omitted, which at the webhook level gets defaulted to `false`.

**Comment by nawazkh** [comment]
> Running a test to check if there's a co-relation between OIDC Issuer configMap being created and Managed Cluster's `OIDCIssuerProfile.Enabled`.
> 
> /hold

Yes. 
Tested it locally. I must have `AzureManagedControlPlane.Spec.oidcIssuerProfile.enabled:true` in order to have OIDC ConfigMap created. 
``` 
❯ kg secrets -A -w
NAMESPACE                           NAME                                              TYPE                            DATA   AGE
.
default                             aks-28954-aso-kubeconfig                          Opaque                          1      7m23s
default                             aks-28954-aso-secret                              Opaque                          4      14m
.
default                             aks-6439-aso-kubeconfig                           Opaque                          1      31m
default                             aks-6439-aso-secret                               Opaque                          4      39m
.
```

```
❯ kg configmaps -A
NAMESPACE                           NAME                                                   DATA   AGE
.
default                             aks-6439-aso-oidc-issuer-profile                       1      31m
.
```

Context: `aks-6439` has  `AzureManagedControlPlane.Spec.oidcIssuerProfile.enabled:true`  and `aks-28954` has  `AzureManagedControlPlane.Spec.oidcIssuerProfile` omitted, which at the webhook level gets defaulted to `false`.

Reference documentation on enabling OIDC for AKS Clusters : https://learn.microsoft.com/en-us/azure/aks/use-oidc-issuer

File: azure/services/managedclusters/spec.go (hunk 5 above)
Comment (nawazkh) [reply azure/services/managedclusters/spec.go:490]: Thanks for pointing it out. I agree that we should move this block down.

Comment (nawazkh) [reply azure/services/managedclusters/spec.go:490]: Webhooks check out, OIDC Issuer cannot be disable once enabled. 
Defaulting webhook: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5e1e3283b38c068797e46fd361b52483717eebf1/api/v1beta1/azuremanagedcontrolplane_default.go#L195-L203

Validating webhook: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5e1e3283b38c068797e46fd361b52483717eebf1/api/v1beta1/azuremanagedcontrolplane_webhook.go#L812-L826

Comment (nawazkh) [reply azure/services/managedclusters/spec.go]: added via : 43d109d495aa3cf36a286b00fc1db95e6157e40a

**Comment by nawazkh** [comment]
> Let's add a brief note in `docs/book/src/topics/managedcluster.md` under the "Security Profile for AKS clusters" heading that when you configure `oidcIssuerProfile.enabled` to true on your `AzureManagedControlPlane` resource, you will have a configmap named "-aso-oidc-issuer-profile" in the same namespace as your `AzureManagedControlPlane` that contains the oidc URL under the `oidc-issuer-profile-url` key.

Added via 43d109d495aa3cf36a286b00fc1db95e6157e40a

**Review by willie-yao — APPROVED** [review]
/lgtm

**Pull request merged**